be/src/exec/sorting/merge_path.cpp

```diff
@@ -0,0 +1,904 @@
+
+        output_size += output_segment->total_len;
+        _output_segments.push_back(std::move(output_segment));
+
+        if (output_size >= _merger->streaming_batch_size()) {
+            break;
+        }
+    }
+}
+
+MergePathCascadeMerger::MergePathCascadeMerger(const int32_t degree_of_parallelism,
+                                               std::vector<ExprContext*> sort_exprs, const SortDescs& sort_descs,
+                                               std::vector<MergePathChunkProvider> chunk_providers,
+                                               const size_t chunk_size)
+        : _chunk_size(chunk_size),
+          _streaming_batch_size(chunk_size * degree_of_parallelism),
+          _degree_of_parallelism(degree_of_parallelism),
+          _sort_exprs(std::move(sort_exprs)),
+          _sort_descs(std::move(sort_descs)),
+          _chunk_providers(std::move(chunk_providers)),
+          _process_cnts(degree_of_parallelism),
+          _output_chunks(degree_of_parallelism) {
+    _working_nodes.resize(_degree_of_parallelism);
+    _metrics.resize(_degree_of_parallelism);
+
+    _forward_stage(detail::Stage::INIT, 1);
+}
+
+bool MergePathCascadeMerger::is_current_stage_finished(int32_t parallel_idx) {
+    if (_is_forwarding_stage.load(std::memory_order_relaxed)) {
+        return true;
+    }
+
+    return parallel_idx >= _working_parallelism ? true
+                                                : _process_cnts[parallel_idx].load(std::memory_order_relaxed) == 0;
+}
+
+bool MergePathCascadeMerger::is_pending(int32_t parallel_idx) {
+    if (_is_forwarding_stage.load(std::memory_order_relaxed)) {
+        return true;
+    }
+
+    if (_stage.load(std::memory_order_relaxed) != detail::Stage::PENDING) {
+        return false;
+    }
+
+    if (is_current_stage_finished(parallel_idx)) {
+        return true;
+    }
+
+    for (auto* leaf : _leafs) {
+        if (leaf->is_pending()) {
+            return true;
+        }
+    }
+
+    if (!_is_first_pending) {
+        const auto now = std::chrono::steady_clock::now();
+        const auto pending_time = std::chrono::duration_cast<std::chrono::nanoseconds>(now - _pending_start).count();
+        std::for_each(_metrics.begin(), _metrics.end(), [pending_time](auto& metrics) {
+            COUNTER_UPDATE(metrics.stage_counter, 1);
+            COUNTER_UPDATE(metrics.stage_counters[detail::Stage::PENDING], 1);
+            COUNTER_UPDATE(metrics.stage_timer, pending_time);
+            COUNTER_UPDATE(metrics.stage_timers[detail::Stage::PENDING], pending_time);
+        });
+    }
+    _is_first_pending = false;
+
+    _forward_stage(detail::Stage::PREPARE, 1);
+    return false;
+}
+
+bool MergePathCascadeMerger::is_finished() {
+    if (_is_forwarding_stage.load(std::memory_order_relaxed)) {
+        return false;
+    }
+
+    return _stage == detail::Stage::FINISHED;
+}
+
+ChunkPtr MergePathCascadeMerger::try_get_next(const int32_t parallel_idx) {
```

Comment:
A stupid question.
`try_get_next ` is the kernel compute code. I see each PallallelSortSourceOperator will trigger it. 

So the sort dop is the same as the sort source operator's dop which is the result of why is called the parallel sort.


Why you say only the operator with driver_sequence = 0 will output data.

Reply:
If all the source operator can output data, extra ordering machenism is needed to re-build a final sorted run.

And no heavy process in the `FETCH_CHUNK` stage, so the performance is acceptable I guess.